### PR TITLE
TMDM-13545 [REST] GET /data/{containerName}/{type}/{id} with datetime fails

### PR DESCRIPTION
**Pull request by zfkang-talend**
Jira: https://jira.talendforge.org/browse/TMDM-13545

**What is the current behavior?** (You should also link to an open issue here)
1. for below datamodel
```
A
|__B (container type)
     |__C
```
If B has a subtype, and C is in the subtype, it doesn't get the field when using B.getField(C).
2. One Field, is one to many, after fetching it into DataRecord, tt's value type is PersistentList which states in Hibernate is persistent, if this field value has no change when move one date time, the new DataRecord for this field's type is not changed, one persistent state data doesn't save when saved, throws
```
Illegal attempt to associate a collection with two open sessions
```
3. For the attachment data model, field detail used inherit, do below operation:

create one record like 
edit this record and remove second detail, saved
Check the UpdateRepord data:
```
<Item>
    <path>detail[2]/code</path>
    <oldValue>bb</oldValue>
  </Item>
  <Item>
    <path>detail[2]/features/actor</path>
    <oldValue>bb-1</oldValue>
  </Item>
  <Item>
    <path>detail[2]/features/vendor[1]</path>
    <oldValue>bb-2</oldValue>
  </Item>
  <Item>
    <path>detail[2]/features</path>
  </Item>
```
for detail[2], only point out the index, but have no type information, if restore detail[2],  it doesn't correct match to ContractDetailSubType.
4. Fetch it from TemporaryStorage and close HistoryStorage
read DataRecord and generate the JSON stream sent to client

for above steps, after fetching the record and closed the HistoryStorage, when generating the JSON, if the field is one-to-many, it will use the lazy strategy to fetch, but now, HistoryStorage had been closed, it throws the Exception:
```
org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: org.talend.mdm.storage.hibernate.X_ANONYMOUS1.x_vendor, could not initialize proxy - no Session
```

**What is the new behavior?**
Solution for above issue:
1. for the inherited type, add the new function to get the fieldType:
```
public static FieldMetadata getFieldMetadata(ComplexTypeMetadata complexTypeMetadata, String element) {
        FieldMetadata field = null;
        if (complexTypeMetadata.hasField(element)) {
            field =complexTypeMetadata.getField(element);
        } else if (complexTypeMetadata instanceof ContainedComplexTypeMetadata) {
            ComplexTypeMetadata contain = ((ContainedComplexTypeMetadata) complexTypeMetadata).getContainedType();
            for (ComplexTypeMetadata complexType : contain.getSubTypes()) {
                for (FieldMetadata subField : complexType.getFields()) {
                    if (field == null && subField.getName().equals(element)) {
                        field = subField;
                        break;
                    }
                }
            }
        }
        return field;
    }
```
2. need transfer PersistentList type to LinkedList type
```
void transformPersistListToList(DataRecord dataRecord) {
        for (FieldMetadata field : dataRecord.getSetFields()) {
            Object value = dataRecord.get(field);
            if (field.isMany() && ((value instanceof List) && !(value instanceof LinkedList) && !(value instanceof ArrayList))
                    && (field instanceof SimpleTypeFieldMetadata)) {
                LinkedList newValue = new LinkedList();
                Iterator iterator = ((List) value).iterator();
                while (iterator.hasNext()) {
                    newValue.add(iterator.next());
                }
                dataRecord.setListValue(field, newValue);
            }
            if (field instanceof ContainedTypeFieldMetadata) {
                if (field.isMany()) {
                    List<DataRecord> listValue = (List) value;
                    for (DataRecord dataRecord1 : listValue) {
                        transformPersistListToList(dataRecord1);
                    }
                } else {
                    transformPersistListToList((DataRecord) value);
                }
            }
        }
    }
```
3. if the action path contains '@xsi:type', when create need specific the new type:
reference code in DataRecordAccessor.java
```
public void createAndSet(String value) {
    if (path.indexOf('@') > 0) {
        create(value);
    } else {
        create(null);
    }
    pathElements = null;
    initPath();
    set(value);
}
```
when create new record, need specific the new type:
reference code in DataRecordAccessor.kava
```
public void create(String specifiedType) {
......
DataRecord record = null;
if (specifiedType != null) {
	ComplexTypeMetadata complexTypeMetadata = ((ContainedTypeFieldMetadata) field).getContainedType().getSubTypes().stream().filter(item -> item.getName().equals(specifiedType)).findFirst().get();
	record = new DataRecord(complexTypeMetadata, UnsupportedDataRecordMetadata.INSTANCE);
} else {
	record = new DataRecord((ComplexTypeMetadata) field.getType(),UnsupportedDataRecordMetadata.INSTANCE);
}
......
}
```
when restore the data, for changeType action, when perform this action, will ignore it
reference code in CompositeAction.java
```
public MutableDocument perform(MutableDocument document) {
 MutableDocument mutableDocument = document;
 List<Action> reorderedAction = reorderDeleteContainedTypeActions(actions);
 for (Action action : reorderedAction) {
	FieldUpdateAction fieldUpdateAction = (FieldUpdateAction) action;
	if (fieldUpdateAction.getPath().indexOf('@') > 0 && fieldUpdateAction.getNewValue() == null
			&& fieldUpdateAction.getOldValue() != null) {
		continue;
	}
	 mutableDocument = action.perform(document);
 }
 return mutableDocument;
}
```
4. for MemoryStorage, the storage name is a UUID, when generating the model.hibernate.xml file, if the storage is MemoryStorage, for the one-to-many field, the lazy *set to *false *from *extra

reference code in DefaultStorageClassLoader.java transfer the value fromMemoryStorage when initilize the class MappingGenerator
```
public MappingGenerator getMappingGenerator(Document document, TableResolver resolver) {
	boolean fromMemoryStorage= false;
	try {
		UUID formatStorageName = UUID.fromString(storageName);
		fromMemoryStorage = true;
	} catch (Exception e) {
		LOGGER.debug("Storage is not one Memory Storage");
	}
	switch (type) {
		case MASTER:
		case SYSTEM:
			return new MappingGenerator(document, resolver, dataSource, true, fromMemoryStorage);
		case STAGING:
			return new MappingGenerator(document, resolver, dataSource, false);
		default:
			throw new NotImplementedException("No support for storage type '" + type + "'.");
	}
}
```
reference code in MappingGenerator.java to modify the laze strategy

```
if (fromMemoryStorage) {
	lazyAttribute.setValue("false"); 
} else {
	lazyAttribute.setValue("extra");
}
```

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java [hunk 1]
```diff
@@ -292,6 +292,32 @@ public void set(FieldMetadata field, Object o) {
...
+     */
+    public void setListValue(FieldMetadata field, Object o) {
+        if (field == null) {
+            throw new IllegalArgumentException("Field can not be null.");
+        }
+        if (!field.isMany()) {
+            fieldToValue.put(field, o);
+        } else {
+            List list = (List) fieldToValue.get(field);
+            if (list == null && o instanceof List) {
```
Comment (hw-zhu) [inline org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: I see your declare type is LinkedList, needn't check it with "o instanceof LinkedList"

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java [hunk 2]
```diff
@@ -135,10 +136,17 @@ protected String getCatalog() {
     }
 
     public MappingGenerator getMappingGenerator(Document document, TableResolver resolver) {
+        boolean fromMemoryStorage = false;
+        try {
+            UUID formatStorageName = UUID.fromString(storageName);
+            fromMemoryStorage = true;
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java]: I suggest using AtomicBoolean replace this one.

File: org.talend.mdm.core/src/com/amalto/core/history/accessor/record/DataRecordAccessor.java [hunk 3]
```diff
@@ -558,4 +577,22 @@ public int compareTo(Accessor accessor) {
...
+        FieldMetadata field = null;
+        if (complexTypeMetadata.hasField(element)) {
+            field = complexTypeMetadata.getField(element);
+        } else if (complexTypeMetadata instanceof ContainedComplexTypeMetadata) {
+            ComplexTypeMetadata contain = ((ContainedComplexTypeMetadata) complexTypeMetadata).getContainedType();
+            for (ComplexTypeMetadata complexType : contain.getSubTypes()) {
+                for (FieldMetadata subField : complexType.getFields()) {
+                    if (field == null && subField.getName().equals(element)) {
+                        field = subField;
+                        break;
```
Comment (hw-zhu) [inline org.talend.mdm.core/src/com/amalto/core/history/accessor/record/DataRecordAccessor.java]: the break only may jump out of the internal For loop, don't outside that one.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java [hunk 4]
```diff
@@ -135,10 +136,17 @@ protected String getCatalog() {
...
+        } catch (Exception e) {
+            LOGGER.debug("Storage is not one Memory Storage");
```
Comment (lpw82gf) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java]:
```
if (LOGGER.isDebugEnabled()) {
    LOGGER.debug("Storage is not one Memory Storage", e);
}
```

Comment (hw-zhu) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java]: why is it a debug level in catch block?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java (hunk 2 above)
Comment (zfkang-talend) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java]: It happens when deploy the data model, it doesn't use the concurrent.

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java [hunk 5]
```diff
@@ -292,6 +292,32 @@ public void set(FieldMetadata field, Object o) {
         }
     }
 
+    /**
+     * if the value is list, set the value to the new value, can't add to original
+     * @param field
+     * @param o
+     */
+    public void setListValue(FieldMetadata field, Object o) {
```
Comment (lpw82gf) [inline org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: Logic of this method is weird, what are we going to do?

File: org.talend.mdm.core/src/com/amalto/core/history/accessor/record/DataRecordAccessor.java [hunk 6]
```diff
@@ -53,7 +54,7 @@ private void initPath() {
 
     @SuppressWarnings("rawtypes")
     private static LinkedList<PathElement> getPath(DataRecord dataRecord, String path) {
-        LinkedList<PathElement> elements = new LinkedList<PathElement>();
+        LinkedList<PathElement> elements = new LinkedList<>();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/history/accessor/record/DataRecordAccessor.java]: `List<PathElement> elements = new LinkedList<>();`

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java [hunk 7]
```diff
@@ -135,10 +136,19 @@ protected String getCatalog() {
     }
 
     public MappingGenerator getMappingGenerator(Document document, TableResolver resolver) {
+        boolean fromMemoryStorage = false;
+        try {
+            UUID formatStorageName = UUID.fromString(storageName);
+            fromMemoryStorage = true;
+        } catch (Exception e) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Storage" + storageName + " is not Memory Storage", e);
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java]: LOGGER.debug("Storage " + storageName + " is not Memory Storage.", e);

File: org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java [hunk 8]
```diff
@@ -292,6 +292,20 @@ public void set(FieldMetadata field, Object o) {
         }
     }
 
+    /**
+     * if the value is list, set the value to the new value, can't add to original
```
Comment (lpw82gf) [inline org.talend.mdm.query/src/com/amalto/core/storage/record/DataRecord.java]: `If value is list, set value to new list, not add to old list`

**Pull request merged**